tools/dependency/validate.py

```diff
@@ -0,0 +1,260 @@
+        '//source/common/crypto/...', '//source/common/conn_pool/...',
+        '//source/common/formatter/...', '//source/common/http/...', '//source/common/ssl/...',
+        '//source/common/tcp/...', '//source/common/tcp_proxy/...', '//source/common/network/...')
+    expected_dataplane_core_deps = self._dep_info.DepsByUseCategory('dataplane_core')
+    bad_dataplane_core_deps = queried_dataplane_core_min_deps.difference(
+        expected_dataplane_core_deps)
+    if len(bad_dataplane_core_deps) > 0:
+      raise DependencyError(
+          f'Observed dataplane core deps {queried_dataplane_core_min_deps} is not covered by '
+          '"use_category" implied core deps {expected_dataplane_core_deps}: {bad_dataplane_core_deps} '
+          'are missing')
+
+  def ValidateControlPlaneDeps(self):
+    """Validate controlplane dependencies.
+
+    Check that we at least tag as controlplane dependencies that match some well-known targets for
+    the control-plane.
+
+    Raises:
+      DependencyError: on a dependency validation error.
+    """
+    print('Validating control-plane dependencies...')
+    # Necessary but not sufficient for controlplane. With some refactoring we could
+    # probably have more precise tagging of dataplane/controlplane/other deps in
+    # these paths.
+    queried_controlplane_core_min_deps = self._build_graph.QueryExternalDeps(
+        '//source/common/config/...')
+    expected_controlplane_core_deps = self._dep_info.DepsByUseCategory('controlplane')
+    bad_controlplane_core_deps = queried_controlplane_core_min_deps.difference(
+        expected_controlplane_core_deps)
+    if len(bad_controlplane_core_deps) > 0:
+      raise DependencyError(
+          f'Observed controlplane core deps {queried_controlplane_core_min_deps} is not covered '
+          'by "use_category" implied core deps {expected_controlplane_core_deps}: '
+          '{bad_controlplane_core_deps} are missing')
+
+  def ValidateExtensionDeps(self, name, target):
+    """Validate that extensions are correctly declared for dataplane_ext and observability_ext.
+
+    Args:
+      name: extension name.
+      target: extension Bazel target.
+
+    Raises:
+      DependencyError: on a dependency validation error.
+    """
+    print(f'Validating extension {name} dependencies...')
+    queried_deps = self._build_graph.QueryExternalDeps(target)
+    marginal_deps = queried_deps.difference(self._queried_core_deps)
+    expected_deps = []
+    for d in marginal_deps:
+      # TODO(htuch): Ensure that queried deps are fully contained in
+      # repository_locations, i.e. that we're tracking with metadata all actual
+      # dependencies. Today, we are missing API and pip3 deps based on manual
+      # inspection.
+      metadata = self._dep_info.GetMetadata(d)
+      if metadata:
+        use_category = metadata['use_category']
+        if ('dataplane_ext' not in use_category and 'observability_ext' not in use_category and
```

Comment:
i love pep8 with the exception of `W503` and `W504` - i find conditions and line breaking like this very hard to grok

i would put these conditions into a variable - ie:
```
thing_i_care_about = (
  some_condition
  and not other_condition)
if thing_i_care_about:
  ...
```
with conjunctions split to the beginning of lines